qiskit_machine_learning/kernels/quantum_kernel.py

```diff
@@ -441,188 +444,224 @@ def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray = None) -> np.ndarray:
                 - x_vec and/or y_vec have incompatible dimension with feature map and
                     and feature map can not be modified to match.
         """
-        # Ensure all training parameters have been bound in the feature map circuit.
-        unbound_params = self.get_unbound_training_parameters()
-        if unbound_params:
-            raise ValueError(
-                f"""
-                The feature map circuit contains unbound training parameters ({unbound_params}).
-                All training parameters must be bound to numerical values before evaluating
-                the kernel matrix.
-                """
-            )
+        self._check_training_parameters_bound()
 
         if self._quantum_instance is None:
             raise QiskitMachineLearningError(
                 "A QuantumInstance or Backend must be supplied to evaluate a quantum kernel."
             )
-        if isinstance(self._quantum_instance, Backend):
-            self._quantum_instance = QuantumInstance(self._quantum_instance)
 
-        if not isinstance(x_vec, np.ndarray):
-            x_vec = np.asarray(x_vec)
-        if y_vec is not None and not isinstance(y_vec, np.ndarray):
-            y_vec = np.asarray(y_vec)
+        x_vec, y_vec = self._validate_input(x_vec, y_vec)
+
+        # determine if calculating self inner product
+        is_symmetric = True
+        if y_vec is None:
+            y_vec = x_vec
+        elif not np.array_equal(x_vec, y_vec):
+            is_symmetric = False
+
+        # initialize kernel matrix
+        kernel = np.zeros((x_vec.shape[0], y_vec.shape[0]))
+
+        # set diagonal to 1 if symmetric
+        if is_symmetric:
+            np.fill_diagonal(kernel, 1)
+
+        is_statevector_sim = self._quantum_instance.is_statevector
+        # calculate kernel
+        if is_statevector_sim:
+            kernel = self._calculate_kernel_statevector(x_vec, y_vec, is_symmetric, kernel)
+        else:
+            kernel = self._calculate_kernel(x_vec, y_vec, is_symmetric, kernel)
+
+        return kernel
+
+    def _validate_input(self, x_vec: np.ndarray, y_vec: np.ndarray):
```

Comment:
Validate might suggest that vectors are not modified in the method but they might be.